scripts/depalign.js

```diff
@@ -0,0 +1,191 @@
+const childProcess = require('child_process');
+const path = require('path');
+const fsExtra = require('fs-extra');
+const semver = require('semver');
+const chalk = require('chalk');
+
+const depalignrc = require('../.depalignrc.json');
+
+const USAGE = `Check for dependency alignment issues.
+
+USAGE: depalign.js [--skip-deduped|--json|--fix]
+
+Options:
+
+--skip-deduped  don't output warnings for ranges that can be resolve to a single version.
+--json          output a json report
+--suggest-fixes output a list of replacements to normalize ranges and align everything
+                to the highest possible range
+
+.depalignrc.json
+
+.depalignrc.json can list ignored dependency ranges that will be excluded from the check.
+
+For example: {"ignore": {"async": ["^1.2.3"]}}.
+`;
+
+const LERNA_BIN = path.resolve(
+  __dirname,
+  '..',
+  'node_modules',
+  '.bin',
+  'lerna'
+);
+
+const ROOT_PACKAGE_JSON_PATH = path.dirname(require.resolve('../package.json'));
+
+function filterOutStarDeps(entries) {
+  return entries.filter(([, v]) => v !== '*');
+}
+
+function collectDependencies() {
+  const packages = JSON.parse(childProcess.execSync(`${LERNA_BIN} list --all --json --toposort`));
+  const dependencies = {};
+
+  for (const packageInfo of [...packages, {location: ROOT_PACKAGE_JSON_PATH}]) {
+    const packageJsonPath = path.join(packageInfo.location, 'package.json');
+    const packageJson = fsExtra.readJSONSync(packageJsonPath);
+    for (const [dependency, versionRange] of [
+      ...Object.entries(packageJson.dependencies || {}),
+      ...Object.entries(packageJson.devDependencies || {}),
+      ...filterOutStarDeps(Object.entries(packageJson.peerDependencies || {})),
+      ...filterOutStarDeps(Object.entries(packageJson.optionalDependencies || {}))
+    ]) {
+      dependencies[dependency] = dependencies[dependency] || [];
+      dependencies[dependency].push(versionRange);
+    }
+  }
+
+  for (const [dependency, versionRanges] of Object.entries(dependencies)) {
+    dependencies[dependency] = Array.from(new Set(versionRanges));
+  }
+
+  return dependencies;
+}
+
+function intersects(range) {
+  for (const v1 of range) {
+    for (const v2 of range) {
+      try {
+        if (!semver.intersects(v1, v2)) {
+          return false;
+        }
+      } catch (e) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
```

Comment:
On one hand my eye got caught on the amount of unnecessary iterations here due to the nested loops, on the other slicing array inside the loop is probably adding the same amount of overhead, so you decide if you want to apply the suggestion 😆  

```suggestion
function intersects(range) {
  for (const [idx, v1] of Object.entries(range)) {
    for (const v2 of range.slice(Number(idx) + 1)) {
      try {
        if (!semver.intersects(v1, v2)) {
          return false;
        }
      } catch (e) {
        return false;
      }
    }
  }
  return true;
}
```

Reply:
Yeah nice, of course i'll commit it :)! Was O(n^2) for small numbers but it did check the same intersection twice. This looks great, I like the `Object.entries(range)` trick!